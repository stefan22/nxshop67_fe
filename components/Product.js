import Link from 'next/link'
import ItemSl from './styles/ItemSl'
import PriceTag from './styles/PriceTag'
import TitleSl from './styles/TitleSl'
import OnSaleSl from './styles/OnSaleSl'
import { GBPCurrencyFormat as poundFormat } from '../lib/currencyFormat'

const Product = ({ product }) => {
  return (
    <ItemSl>
      <h2>{product?.name}</h2>
      <h3 className="heading_H3">{product?.description}</h3>
      <OnSaleSl>ON-SALE!</OnSaleSl>
      <picture>
        <source srcSet={product?.photo?.image.publicUrlTransformed} />
        <img
          src={product?.photo?.image.publicUrlTransformed}
          alt={product?.name}
        />
      </picture>

      <TitleSl>
        <Link href={`/product/${product?.id}`}>{product?.name}</Link>
      </TitleSl>
      <PriceTag>{poundFormat(product?.price)}</PriceTag>
    </ItemSl>
  )
}

export default Product
