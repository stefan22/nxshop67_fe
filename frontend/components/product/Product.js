import React from 'react'
import Link from 'next/link'
import Image from 'next/image'
import {
  TitleContainer,
  OnSaleContainer,
  PriceTag,
  ItemContainer
} from './Product.styles'
import { GBPCurrencyFormat as poundFormat } from '../../utils/currencyFormat'
import { DeleteProduct } from '../delete-product'

const Product = ({ product }) => (
  <ItemContainer animate={{ opacity: 1 }} data-testid="product">
    <OnSaleContainer>ON-SALE!</OnSaleContainer>
    <Link
      passHref
      href={{
        query: { id: `${product?.id}` },
        pathname: `/product/${product?.id}`
      }}
    >
      <picture>
        <Image
          layout="intrinsic"
          objectFit="cover"
          loading="lazy"
          width={267}
          height={294}
          src={product?.photo?.image.publicUrlTransformed}
          alt={product?.name}
        />
      </picture>
    </Link>
    <PriceTag>{poundFormat(product?.price)}</PriceTag>
    <TitleContainer>
      <span>{product?.name}</span>

      <Link
        passHref
        style={{ color: '#c28f87' }}
        href={{
          pathname: '/update',
          query: {
            id: product?.id
          }
        }}
      >
        Edit &#x23F5;
      </Link>

      <DeleteProduct btnName="Delete" id={product?.id} />
    </TitleContainer>
  </ItemContainer>
)

export default Product
