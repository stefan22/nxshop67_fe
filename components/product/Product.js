import React from 'react'
import Link from 'next/link'
import Image from 'next/image'
import { TitleContainer, OnSaleContainer, PriceTag, ItemContainer } from './Product.styles'
import { GBPCurrencyFormat as poundFormat } from '@/lib/currencyFormat'

const Product = ({ product }) => (
  <ItemContainer data-testid="product">
    <OnSaleContainer>ON-SALE!</OnSaleContainer>
    <Link href={`/product/${product?.id}`}>
      <picture>
        <Image
          layout="intrinsic"
          objectFit="cover"
          loading="lazy"
          width={267}
          height={294}
          src={product?.photo?.image.publicUrlTransformed}
          alt={product?.name}
        />
      </picture>
    </Link>
    <PriceTag>{poundFormat(product?.price)}</PriceTag>
    <TitleContainer>
      <span>{product?.name}</span>

      <Link
        href={{
          pathname: '/update',
          query: {
            id: product?.id
          }
        }}
      >
        Edit &#x23F5;
      </Link>
    </TitleContainer>
  </ItemContainer>
)

export default Product
