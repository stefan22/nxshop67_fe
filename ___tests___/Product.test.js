/**
 * @jest-environment jsdom
 */

import '@testing-library/jest-dom'
import { render, screen } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import Product from '../components/Product'

const singleProduct = () => ({
    id: "abc123",
    price: "Â£7.70",
    user: null,
    name: "ABC",
    description: "Very nice hat",
    photo: {
        id:"abc123",
        altText: "hat item",
        image: {
            publicUrlTransformed: "item.jpg"
        }
    }
})

const product = singleProduct()

describe("<Product />", () => {

    it("should render product name in page", () => {
        const { container, debug } = render(
          <MockedProvider>
              <Product product={product} />
          </MockedProvider>
        )
        const expectedProductName = screen.getAllByText('ABC')[0]
        expect(expectedProductName).toBeInTheDocument()
    })

    it("should render product description in page", () => {
        const { container, debug } = render(
          <MockedProvider>
              <Product product={product} />
          </MockedProvider>
        )
        const desc = screen.getByText('Very nice hat')
        expect(desc).toBeInTheDocument()
    })

    it("should have an <a /> tag with atributes", () => {
        const { container, debug } = render(
          <MockedProvider>
              <Product product={product} />
          </MockedProvider>
        )
        const linkTag = container.querySelector('a');
        expect(linkTag).toHaveAttribute('href','/product/abc123')
        expect(linkTag).toHaveTextContent(product.name)
    })

    it("should have an image with src and alt attributes", () => {
        const { container, debug } = render(
          <MockedProvider>
              <Product product={product} />
          </MockedProvider>
        )
        const picImg = container.querySelector('picture > img')
        expect(picImg).toHaveAttribute('src','item.jpg')
        expect(picImg).toHaveAttribute('alt','ABC')
    })

    it('product container matches the snapshot', () => {
        const { container } = render(
          <MockedProvider>
              <Product product={product} />
          </MockedProvider>
        )
        expect(container).toMatchSnapshot();
    })
})
