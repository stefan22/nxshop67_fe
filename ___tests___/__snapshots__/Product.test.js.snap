// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Product /> product container matches the snapshot 1`] = `
<div>
  <div
    class="ItemSl__Item-sc-1s91ler-0 gVQryV"
    data-testid="product"
  >
    <h2>
      ABC
    </h2>
    <h3
      class="heading_H3"
    >
      Very nice hat
    </h3>
    <span
      class="OnSaleSl-sc-17gb5te-0 fyniQW"
    >
      ON-SALE!
    </span>
    <picture>
      <source
        srcset="item.jpg"
      />
      <img
        alt="ABC"
        src="item.jpg"
      />
    </picture>
    <h3
      class="TitleSl-sc-fwsgza-0 kFZlbu"
    >
      <a
        href="/product/abc123"
      >
        ABC
      </a>
    </h3>
    <span
      class="PriceTag-sc-vp1xmv-0 dpkGEc"
    >
      Amount passed not a number
    </span>
  </div>
</div>
`;
